### ISSUE TITLE
Private artifactory server credentials get passed on to upstream unauthenticated chart repository

### ISSUE THREAD

**Issue by olhado** [helm/helm#9777]
Output of `helm version`:

`version.BuildInfo{Version:"v3.6.0", GitCommit:"7f2df6467771a75f5646b7f12afb408590ed1755", GitTreeState:"dirty", GoVersion:"go1.16.4"}`

Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T21:16:14Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.8-eks-96780e", GitCommit:"96780e1b30acbf0a52c38b6030d7853e575bcdf3", GitTreeState:"clean", BuildDate:"2021-03-10T21:32:29Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  EKS

---
We have a setup where we are trying to use a virtual artifactory helm repository to proxy to various third-party chart repositories. Our artifactory system is private, and requires credentials. What we have found is that it appears helm always passes along the credentials to every HTTP request in the install command. This leads to an error with at least one repository based in S3 where AWS responds with a `400: Bad Request` response. We do not see this issue if we install directly from originating chart repository.

Does it make sense to assume the credentials passed to get the manifest of charts should be continued to be used if the URL is different? The specific lines of code helm is erroring at are https://github.com/helm/helm/blob/main/pkg/getter/httpgetter.go#L59-L61

**Comment by bacongobbler** [comment]
This is a very good point. One that I was just thinking about last week while thinking about some of the concerns #9762 raises. Thank you for writing this up.

I think there is a general security concern for helm to assume that the credentials used to authenticate with the chart repository API can also be used to fetch the chart itself using the download URL from the index.

> Does it make sense to assume the credentials passed to get the manifest of charts should be continued to be used if the URL is different?

In many cases yes. Many organizations may choose to have both the chart repository API and the charts themselves backed by the same authentication endpoints. For example if both are saved in an Azure account and you want them to share the same [service principal](https://docs.microsoft.com/en-us/azure/active-directory/develop/app-objects-and-service-principals) as they are part of the same "app" or "platform" you're building.

However the case you mention (a chart index pointing at external endpoints, like a public mirror) could be a privacy issue for many users.

An open question here: if both the chart repository and the download URL are saved at different locations but they share the same credentials, how can one pass those credentials onto the secondary fetch? If we solve this problem, then we can start to discuss how to pull apart the repository API from the download URL.

That all being said, the [experimental OCI feature](https://helm.sh/docs/topics/registries/) is expected to replace the chart repository API moving forward, so it might not be worth spending too much time trying to solve this problem if it's going to be deprecated once that goes GA.

**Comment by olhado** [comment]
I see, makes sense. My concern is the ability of vendors like Artifactory updating their software to support OCI quickly, or seamlessly. Also, looking into OCI, it doesn't seem to address a proxy/cache situation? At least, not in the way tools like Artifactory, Nexus, and others often implement aggregating remote and local (as in created in that repository system itself) repositories mostly as pass-throughs.

Could a patch that adds a flag to skip that logic block be an acceptable short-term patch? Where by default it does the current behavior?

**Comment by bacongobbler** [comment]
> Also, looking into OCI, it doesn't seem to address a proxy/cache situation? At least, not in the way tools like Artifactory, Nexus, and others often implement aggregating remote and local (as in created in that repository system itself) repositories mostly as pass-throughs.

It's described [in the documentation](https://docs.docker.com/registry/recipes/mirror/).

**Comment by bacongobbler** [comment]
fixed in Helm 3.6.1. See https://github.com/helm/helm/security/advisories/GHSA-56hp-xqp3-w2jf.

Thanks for raising this issue.
